### ISSUE TITLE
Potentially private IP addresses revealed

### ISSUE THREAD

**Issue by stefhak** [w3c/webrtc-pc#179]
Host IP addresses are made available to an application that creates a PeerConnection and lets it start gathering ICE candidates, and this can be done without the user knowing.
This can be sensitive in certain scenarios (e.g. VPN) as recently discussed, OTOH host IP addresses (if reachable from a peer) are the "best" for establishing connectivity.
This issue is opened to keep track of any decisions made regarding this problem.

**Comment by martinthomson** [comment]
it's not sufficient to disable host addresses.  If you care about fingerprinting, then reflexive addresses can be problematic as well.

**Comment by juberti** [comment]
Disabling host addresses will require use of a STUN server, as well as a NAT that supports hairpinning, to allow connections between peers on the same LAN.

**Comment by martinthomson** [comment]
If we consider this to have the same requirements as drive the [ICE transport policy](http://w3c.github.io/webrtc-pc/#idl-def-RTCIceTransportPolicy), then relay candidates are the only ones that should be permitted.

That said, I think that addressing this concern is not the business of the specification.  The specification need only note what information is exposed and describe the consequences of that exposure.

**Comment by stefhak** [comment]
(@martinthomson: about the reflexive addresses and fingerprinting: are these not made available to the site already when the page is fetched - this is where the server sends the document?)

**Comment by martinthomson** [comment]
@stefhak, the concerns that I've seen raised are with respect to users who are behind a proxy or VPN.  Presumably, the reflexive address that is revealed as a result is different to one that might be revealed by a STUN request made from a host candidate.

**Comment by alvestrand** [comment]
Two different concerns:
- Local IP addresses are revealed to the Web page (we can hide them from correspondents) - but these are vital for efficient local connections. Fingerprint/benefit tradeoff. I suggest we document these and move on.
- Public IP addresses other than the one used as a source for browsing are revealed. This appears to depend on specific VPN configuration properties. More investigation needed.

**Comment by dontcallmedom** [comment]
To add another angle, these IP addresses are potentially revealed to the server, but also to the peer.

[WebRTC Security Architecture](https://tools.ietf.org/html/draft-ietf-rtcweb-security-arch-10#section-5.4) sets requirements on the API to help avoid leaking IP addresses to the peer — we probably should review whether these requirements are currently addressed.

It also says protecting from the server is not a requirement; we probably would want to import some of the language there in the security considerations.

(I also think we need to do more than just saying it's out of scope for the spec — but we need a good sense of the problem space before we can address it effectively)

**Comment by alvestrand** [comment]
Re leaking to the peer: The current controls (RTCIceTransportPolicy) serve exactly this purpose; the policy "relay" will make sure IP addresses are not leaked to the peer; the peer will only see the address of the TURN relay.

We do need to give enough information in the spec to let the implementors do proper design tradeoffs.

**Comment by stefhak** [comment]
Regarding leaking to the peer: I have the same conclusion as you have.

Unfortunately it only helps the user if the app is well behaved (and that
app could equally well have discarded those candidates - the API is just
convenient).

On Fri, Feb 13, 2015 at 4:26 PM, Harald Alvestrand <notifications@github.com

**Comment by Uranium235** [comment]
alvestrand wrote:

>Public IP addresses other than the one used as a source for browsing are revealed. This appears to >depend on specific VPN configuration properties. More investigation needed.

I have experienced the Problem with OpenVPN. The OpenVPN How-To Guide instructs to use the following server configuration:

push "redirect-gateway def1 bypass-dhcp"

The problem is the use of the "def1" flag. It has the effect that the client does not replace the local default route, but overrides it by routing all internet traffic through the tunnel for the more specific routes 0.0.0.0/1 and 128.0.0.0/1 rather than 0.0.0.0/0.

The effect with WebRTC is, that both external IP addresses are discovered. The one from the still existing local default route and the one from the VPN connection.

It has serious privacy implications when remote websites are able to discover the fact that the user is likely to use a VPN service and even more so if the "real" IP address (and thus location) of VPN users can be revealed.

I don't know why this is happening. But if OpenVPN is configured without the "def1" flag and thus removes the existing default route from the system for the duration of the VPN connection, the IP does not get discovered anymore.

Shouldn't the system's routing table be trusted and not being tried to discover every possible path?

**Comment by juberti** [comment]
If you are connected to a corp VPN, does that mean that you necessarily want to run your real-time media over said VPN, especially if it is a SSLVPN that tunnels over TCP? Or should two users inside an enterprise be forced to send their media to an external TURN server rather than connecting directly? 

There are multiple reasons for wanting to be able to use a specific NIC, but I agree that users should have the ability to control this behavior, even if it sacrifices QoS.

We are investigating this in Chrome right now.

**Comment by Uranium235** [comment]
My post above was my experience with my own private VPN that is intentionally set up to route *all* traffic through the tunnel (to hide my real location and/or when using untrusted networks/hotspots).

My corp VPN on the other hand does not set up such a default route, only a route for the corp subnet.

If a corp wants to also route all other traffic through it's systems, that most likely has a reason and would not make them too happy if my browser was able to find an easy way out another way. Not so much a privacy concern then, but probably a corp security concern.

**Comment by juberti** [comment]
For Chrome 42 we added an option to force default routing, which is settable by users and enterprise policy. When this option is set, we will also not generate any candidates with private IPs.

**Comment by stefhak** [comment]
Interesting! Can you elaborate a bit more, e.g. what is default, can it be set per origin (a whitelist to which private IP's are revealed even though this option is set etc.)?

**Comment by juberti** [comment]
Right now it is off by default but can be enabled by the user, a chrome extension, or by policy.

**Comment by dontcallmedom** [comment]
@juberti sent a related proposal to the RTCWeb mailing list: http://www.ietf.org/mail-archive/web/rtcweb/current/msg14494.html

**Comment by fluffy** [comment]
I think the rtcweb WG is right place to resolve this.

**Comment by dontcallmedom** [comment]
The relevant PR was merged so I think this can be closed
